libcudacxx/.upstream-tests/utils/nvidia/nvrtc/nvrtcc.cpp

```diff
@@ -0,0 +1,302 @@
+        std::regex("^-o (.+)[\\\\/]([^\\\\/]+)\\..+$"),
+            [](const std::smatch& match) {
+                outputDir = match[1].str();
+                outputFile = match[2].str();
+                return NORMAL;
+            }
+    },
+    {
+        make_greedy_handler("^-gencode$")
+    },
+    {
+        // Matches '-gencode=' or '-gencode ...'
+        std::regex("^-gencode[= ]?(.+)$"),
+            [](const std::smatch& match) {
+                buildList.emplace(translate_gpu_arch(match[1].str().data()));
+                return NORMAL;
+            }
+    },
+    {
+        // Matches the many various versions of dialect switch and normalizes it
+        std::regex("^[-/]std[:=](.+)$"),
+            [](const std::smatch& match) {
+                nvrtcArguments.emplace_back("-std="+match[1].str());
+                return NORMAL;
+            }
+    },
+    {
+        // Capture an argument that is just '-'. If no input file is listed input is on stdin
+        std::regex("^-$"),
+            [](const std::smatch& match) {
+                inputFile = match[0].str();
+                return NORMAL;
+            }
+    },
+    {
+        // If an input lists a .gpu file, run that file instead
+        std::regex("^([^-].*).gpu$"),
+            [](const std::smatch& match) {
+                execute = true;
+                executionConfig = ExecutionConfig {
+                    RunConfig{1, 0},
+                    {match[0].str()}
+                };
+
+                return NORMAL;
+            }
+    },
+    {
+        // If an input is a .exe file, search for other builds and run those
+        std::regex("^([^-].*).exe$"),
+            [](const std::smatch& match) {
+                execute = true;
+                executionConfig = load_execution_config_from_file(match[1].str() + ".build.yml");
+                assert(executionConfig.builds.size());
+                return NORMAL;
+            }
+    },
+    {
+        // Capture any argument not starting with '-' as the input file
+        std::regex("^([^-].+)[\\\\/].+$"),
+            [](const std::smatch& match) {
+                inputFile = match[0].str();
+                // Capture directory of input file as an include path
+                nvrtcArguments.emplace_back("-I " + match[1].str());
+                return NORMAL;
+            }
+    },
+    {
+        // Throw away remaining arguments
+        std::regex("^-.+$"),
+            [](const std::smatch& match) {
+                ignoredArguments.emplace_back(match[0].str());
+                return NORMAL;
+            }
+    },
+};
+
+int main(int argc, char **argv) {
+    // Greedily take off first arg
+    g_argc = argc-1;
+    g_argv = argv+1;
+
+    ArgProcessorState argState = NORMAL;
+
+    // Start by parsing arguments and building the configuration
+    std::string c_arg{};
+    for (auto a = g_argv; a < g_argv + g_argc; a++) {
```

Comment:
snarky suggestion: `std::regex`. I'm not even mad, I'm impressed. However, would it be easier to just use an existing lib like [cxxopts](https://github.com/jarro2783/cxxopts) for this?

Reply:
I think as we discussed. Pulling in dependencies where we might need to be building in 'weird places' makes me not want to use them... STL-only seems safest from a 'will work almost anywhere' standpoint.